### PR TITLE
Added tests for localisation

### PR DESCRIPTION
**Pull request by shivam71**
Automated tests for testing frontend localisation.

### CODE REVIEW

File: vscode/src/test/suite/testutils.ts [hunk 1]
```diff
@@ -192,3 +192,278 @@ export async function dumpJava(): Promise<void> {
...
+    const regexp = /\{[^\{\}]*\}/g;
+    const params1 = new Set([...str1.matchAll(regexp)].map((value) => value[0]));
+    const params2 = new Set([...str2.matchAll(regexp)].map((value) => value[0]));
+    return setEqual(params1, params2);
+}
+
+function setEqual(setA: Set<string>, setB: Set<string>): boolean {
+    for (const elem of setA) {
+        if (setB.has(elem)) continue;
+        return false;
```
Comment (sid-srini) [inline vscode/src/test/suite/testutils.ts:348]: You may simplify this to
```suggestion
        if (!setB.has(elem)) return false;
```

File: vscode/src/test/suite/testutils.ts [hunk 2]
```diff
@@ -192,3 +192,278 @@ export async function dumpJava(): Promise<void> {
...
+
+/**
+ * 
+ * @param dictString String having key value pairs {key1:value1,key2:value2...}
+ * @returns Set of keys used in the dictString
+ */
+function getPlaceholders(dictString: string): Set<string> {
+    const placeholders = new Set<string>();
+    const cleanedDictString = dictString.replace(/"[^"]*"/g, "SOME_STRING_VALUE").replace(/'[^']*'/g, "SOME_STRING_VALUE");
+    for (const keyVal of cleanedDictString.split(',')) {
```
Comment (sid-srini) [inline vscode/src/test/suite/testutils.ts:375]: Perhaps skip the first element of keyVal since that is expected to be the key?

File: vscode/src/test/suite/testutils.ts [hunk 3]
```diff
@@ -192,3 +192,278 @@ export async function dumpJava(): Promise<void> {
...
+        // improve this so that if value has key:"," that doesn't get picked up
```
Comment (sid-srini) [inline vscode/src/test/suite/testutils.ts:376]: Based on the pre-processing of dictString to replace all quoted strings with a fixed constant, would this case still occur?

File: vscode/src/test/suite/testutils.ts [hunk 4]
```diff
@@ -192,3 +192,278 @@ export async function dumpJava(): Promise<void> {
...
+}
+
+/**
+ * 
+ * @param filePath 
+ * @param pattern To extract the keys and the placeholder map used when calling l10n.value
+ * @param validKeyValues From the bundle.en.json filee 
+ * @returns All usage of l10n.value according to the pattern is having valid keys and placeholder map
+ */
+function checkl10UsageInFile(filePath: string, pattern: RegExp, validKeyValues: any): boolean {
```
Comment (sid-srini) [inline vscode/src/test/suite/testutils.ts:389]: fix typo `checkl10UsageInFile` to `checkL10nUsageInFile`

File: vscode/src/test/suite/testutils.ts [hunk 5]
```diff
@@ -192,3 +192,278 @@ export async function dumpJava(): Promise<void> {
...
+    return result;
+}
+
+/**
+ * 
+ * @param value string to be checked
+ * @returns value of the form %key% where key is some string containing alphanumeric characters 
+ */
+function isLocalizedVal(value: string): boolean {
+    return (value[0] === value[value.length - 1]) && value[0] === '%';
```
Comment (sid-srini) [inline vscode/src/test/suite/testutils.ts:426]:
```suggestion
    return value.length > 2 && value[0] === '%' && value[0] === value[value.length - 1];
```

File: vscode/src/test/suite/testutils.ts [hunk 6]
```diff
@@ -192,3 +192,278 @@ export async function dumpJava(): Promise<void> {
...
+        if (!Array.isArray(fieldVals)) fieldVals = [fieldVals];
+        for (const fieldVal of fieldVals) {
+            localizingKey = getlocalizingKey(fieldVal);
+            if (!isLocalizedVal(fieldVal)) {
+                console.log(`${category} object with id ${id} has a unlocalized field field:'${field}'`);
+                localized = false;
+            } else if (!validKeys.has(localizingKey)) {
+                console.log(`${category} object of id '${id}' has a invalid  localizing key for the field:'${field}' key:'${localizingKey}'`);
+                localized = false;
+            }
```
Comment (sid-srini) [inline vscode/src/test/suite/testutils.ts:465]: To avoid an array index out of bounds err.
```suggestion
            if (!isLocalizedVal(fieldVal)) {
                console.log(`${category} object with id ${id} has a unlocalized field field:'${field}'`);
                localized = false;
            } else {
                localizingKey = getlocalizingKey(fieldVal);
                if (!validKeys.has(localizingKey)) {
                    console.log(`${category} object of id '${id}' has a invalid  localizing key for the field:'${field}' key:'${localizingKey}'`);
                    localized = false;
                }
            }
```

File: vscode/src/test/suite/testutils.ts [hunk 7]
```diff
@@ -192,3 +192,278 @@ export async function dumpJava(): Promise<void> {
...
+            fieldVal = command[field];
+            if (fieldVal === undefined) continue;
+            localizingKey = getlocalizingKey(fieldVal);
+            if (!isLocalizedVal(fieldVal)) {
+                localized = false;
+                console.log(`${field} of the command '${id}' is not being localized`)
+            } else if (!validKeys.has(localizingKey)) {
+                console.log(`Invalid localizing key for the command '${id}' key '${localizingKey}' not found in the package.nls.json`);
+            }
+        }
```
Comment (sid-srini) [inline vscode/src/test/suite/testutils.ts:326]: Could this be refactored to re-use the same code as in `isLocalizedObj()`? That would take care of the array out-of-bounds err mentioned there.

File: vscode/src/test/suite/testutils.ts [hunk 8]
```diff
@@ -192,3 +192,278 @@ export async function dumpJava(): Promise<void> {
...
+        if (!isLocalizedObj(property, propertiesLocalisableFields, propertyId, "Configuration Property", validKeys)) localized = false;
+    }
+    return localized;
+}
+
+export function checkDebuggersLocalisation(debuggers: any, validKeys: Set<string>): boolean {
+    let localized = true;
+    const propertiesLocalisableFields = ['description'];
+    const configLocalisableFields = ['name'];
+    const snippetLocalisableFields = ['label', 'description'];
```
Comment (sid-srini) [inline vscode/src/test/suite/testutils.ts:275]: If available, would you please add the URL of the spec document which defines these as a comment here?

Comment (shivam71) [reply vscode/src/test/suite/testutils.ts:275]: It's not available to my knowledge.

File: vscode/src/test/suite/testutils.ts (hunk 5 above)
Comment (shivam71) [reply vscode/src/test/suite/testutils.ts:426]: Thanks for pointing this out !

File: vscode/src/test/suite/testutils.ts (hunk 7 above)
Comment (shivam71) [reply vscode/src/test/suite/testutils.ts:326]: Sure !

File: vscode/src/test/suite/testutils.ts (hunk 4 above)
Comment (shivam71) [reply vscode/src/test/suite/testutils.ts:389]: Thanks for reading so carefully !

File: vscode/src/test/suite/extension.test.ts [hunk 9]
```diff
@@ -299,4 +299,61 @@ suite('Extension Test Suite', function () {
         assert.ok(!fs.existsSync(mainClass), "Class created by compilation: " + mainClass);
     }).timeout(10000);
 
+     // Check the consistency of the keys and value templates across the bundle files for the supported languages 
+    test("Consistency of keys across bundle.l10n.lang.json files for supported languages", async () => {
+        const extension = extensions.getExtension("oracle.oracle-java");
+        assert(extension);
+        const enBundlePath = path.join(extension.extensionPath, "l10n", "bundle.l10n.en.json");
+        assert.ok(fs.existsSync(enBundlePath), "bundle.l10n.en.json doesn't exists");
+        const supportedLanguages = new Set(["ja", "zh-cn"]);
```
[the remark this answers is not in the document]
Comment (shivam71) [reply vscode/src/test/suite/extension.test.ts:308]: Sure!

File: vscode/src/test/suite/extension.test.ts [hunk 10]
```diff
@@ -30,8 +30,8 @@ import * as myExtension from '../../extension';
 import * as myExplorer from '../../explorer';
 
 import { CodeAction, commands, extensions, Selection, Uri, window, workspace, TreeItem } from 'vscode';
-import { assertWorkspace, dumpJava, getFilePaths, openFile, prepareProject, replaceCode} from './testutils';
 import {SAMPLE_CODE_FORMAT_DOCUMENT, SAMPLE_CODE_SORT_IMPORTS, SAMPLE_CODE_UNUSED_IMPORTS } from './constants';
+import { assertWorkspace, checkCommandsLocalisation, checkConfigurationLocalisation, checkDebuggersLocalisation, checkl10nUsageInFiles, checkViewsLocalisation, dumpJava, getFilePaths, getKeysFromJSON, matchKeys, matchValuesTemplate, openFile, prepareProject, replaceCode } from './testutils';
```
[the remark this answers is not in the document]
Comment (shivam71) [reply vscode/src/test/suite/extension.test.ts:34]: Yes though then we will have to do it as 
`import * as testutils  from './testutils';`
Further everything will be used by doing something  like `testutils.assertWorkspace` . So I  think its okay it being as it is .

File: vscode/src/test/suite/testutils.ts [hunk 11]
```diff
@@ -192,3 +192,278 @@ export async function dumpJava(): Promise<void> {
...
 }
+
+
+export function getKeysFromJSON(filePath: string): Set<string> {
+    return new Set(Object.keys(JSON.parse(fs.readFileSync(filePath, 'utf8'))));
+}
+
+export function matchValuesTemplate(jsonFileBase: string, jsonFileTarget: string): boolean {
+
+    if (!matchKeys(jsonFileBase, jsonFileTarget)) {
```
[the remark this answers is not in the document]
Comment (shivam71) [reply vscode/src/test/suite/testutils.ts:203]: I just wanted to have matchValuesTemplate as a standalone reusable function hence this matchKeys used here .

File: vscode/src/test/suite/testutils.ts (hunk 1 above)
Comment (shivam71) [reply vscode/src/test/suite/testutils.ts:348]: Cool thanks !

File: vscode/src/test/suite/testutils.ts (hunk 3 above)
Comment (shivam71) [reply vscode/src/test/suite/testutils.ts:376]: Nope  I just forgot to remove the comment.

File: vscode/src/test/suite/testutils.ts (hunk 2 above)
Comment (shivam71) [reply vscode/src/test/suite/testutils.ts:375]: Hmm I don't think so, it will be a key value pair only something like `key1:value1` assuming the `cleanedDictString` is  like `"key1:value1 , key2:value2 , key3:value3..."`

**Comment by shivam71** [comment]
@sid-srini  Please review again , post that I will squash and merge after the refactored extension code gets merged . I will need to make some more changes then .

File: vscode/src/test/suite/testutils.ts (hunk 2 above)
Comment (sid-srini) [reply vscode/src/test/suite/testutils.ts:357]: Understood. Thanks.

File: vscode/src/test/suite/testutils.ts (hunk 6 above)
Comment (sid-srini) [reply vscode/src/test/suite/testutils.ts:447]: Sorry, I wrote arr index out of bounds instead of a possible string index err in `getlocalizingKey(fieldVal)`.
However, I just realised that the js `string.substring()` does not throw an err, but instead will return an unexpected output when the `endIndex < beginIndex`. 
So, the change in order of the method invocations - first `isLocalizedVal`, then `getlocalizingKey` will only have a minor performance improvement.

Comment (shivam71) [reply vscode/src/test/suite/testutils.ts:447]: Then I guess we are okay not changing it right ?

**Review by sid-srini — DISMISSED** [review]
Seems fine. Thanks 👍

File: vscode/src/test/testutils.ts [hunk 12]
```diff
@@ -333,42 +333,298 @@ export function runTestSuite(folder: string): Promise<void> {
...
+	if (extension.isActive) {
+		return globalVars.clientPromise.client;
+	}
+	const waitForExtenstionActivation: Thenable<NbLanguageClient> = extension.activate().then(async () => {
+		return await globalVars.clientPromise.client;
+	});
+	return Promise.resolve(waitForExtenstionActivation);
+}
+
+export function findClusters(myPath: string): string[] {
```
Comment (sid-srini) [inline vscode/src/test/testutils.ts:350]: Is this method still required to be added to *testutils.ts*?

Comment (shivam71) [reply vscode/src/test/testutils.ts:350]: Seems like it's not being used anywhere post the refactoring . I am actually not adding this one the diff is  a bit misleading . Nevertheless I feel we can keep this even though we are not using it since we may intend to use it in the future .

File: vscode/src/test/testutils.ts
Comment (sid-srini) [inline vscode/src/test/testutils.ts:1]: Ok. It seems like the linting had converted indents to 8 spaces instead of 4 spaces. That's perhaps why there appear to be unrelated changes. 
The problem seems to be that the functions prior to the modified ones seem to have the incorrect indent of 8 spaces. 
Would you please re-run your indent linting on this file to bring it in line with the project's 4 spaces convention? 

Thanks.

File: vscode/src/test/suite/localisation/index.ts [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { runTestSuite } from '../../testutils';
+
+export function run(): Promise<void> {
+	return runTestSuite(__dirname);
```
Comment (sid-srini) [inline vscode/src/test/suite/localisation/index.ts:25]: See the 4-space indentation comment below for testutils.ts

**Review by sid-srini — CHANGES_REQUESTED** [review]

File: vscode/src/test/testutils.ts
Comment (shivam71) [reply vscode/src/test/testutils.ts:1]: Thanks for finding the issue with the diff !
Seems to be that tabs vs spaces issue I just converted all my tabs to spaces and pushed again now it seems okay .

**Review by sid-srini — APPROVED** [review]
Thanks @shivam71. 👍 LGTM

**Review by Achal1607 — APPROVED** [review]
LGTM. Thanks!
